tests/benchmark/crud_test.go

```diff
@@ -0,0 +1,895 @@
+	if valueExists {
+		return fmt.Sprintf("projects/%s/locations/global", registryProject)
+	}
+	return "projects/bench/locations/global"
+}
+
+func getNumberOfVersions() int {
+	strNumberOfVersions, exists := os.LookupEnv("REGISTRY_VERSION_COUNT")
+	numberOfVersions, err := strconv.Atoi(strNumberOfVersions)
+	if err != nil || !exists {
+		numberOfVersions = 1
+	}
+	return numberOfVersions
+}
+
+func getApiName(i int) string {
+	apiNamePrefix, valueExists := os.LookupEnv("API_NAME_PREFIX")
+	if !valueExists {
+		apiNamePrefix = "benchtest"
+	}
+	apiNameStartOffset, valueExists := os.LookupEnv("API_NAME_START_OFFSET")
+	if valueExists {
+		intVar, _ := strconv.Atoi(apiNameStartOffset)
+		i += intVar
+	}
+
+	return fmt.Sprintf("%s-%d", apiNamePrefix, i)
+}
+
+var rootResource = getRootResource()
+
+func readAndGZipFile(filename string) (*bytes.Buffer, error) {
+	fileBytes, _ := ioutil.ReadFile(filename)
+	var buf bytes.Buffer
+	zw, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	_, err := zw.Write(fileBytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := zw.Close(); err != nil {
+		return nil, err
+	}
+	return &buf, nil
+}
+
+func BenchmarkCreateApi(b *testing.B) {
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	for i := 1; i <= b.N; i++ {
+		if i%1000 == 0 {
+			//Sleep 10s every 1000 iterations to reduce server load
+			time.Sleep(10 * time.Second)
+		}
+
+		apiId := getApiName(i)
+		req := &rpc.CreateApiRequest{
+			Parent: rootResource,
+			ApiId:  apiId,
+			Api: &rpc.Api{
+				DisplayName: apiId,
+				Description: fmt.Sprintf("Description for %s", apiId),
+			},
+		}
+		b.StartTimer()
+		_, err := client.CreateApi(ctx, req)
+		b.StopTimer()
+		/**
+		Go Benchmarking always runs all the benchmark tests once before it runs
+		for the duration specified. So the first row always returns error.
+		*/
+		if i > 1 && err != nil {
+			b.Errorf("CreateApi(%+v) returned unexpected error: %s", req, err)
+		}
+	}
+}
+
+func BenchmarkCreateApiVersion(b *testing.B) {
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	b.StopTimer()
+	b.ResetTimer()
+
+	for i := 1; i <= b.N; i++ {
+		if i%1000 == 0 {
+			//Sleep 10s every 1000 iterations to reduce server load
+			time.Sleep(10 * time.Second)
+		}
+
+		apiId := getApiName(i)
+		for j := 1; j <= getNumberOfVersions(); j++ {
+			req := &rpc.CreateApiVersionRequest{
+				Parent: fmt.Sprintf("%s/apis/%s", rootResource, apiId),
+				ApiVersion: &rpc.ApiVersion{
+					DisplayName: fmt.Sprintf("v%d", j),
+					Description: fmt.Sprintf("v%d of %s", j, apiId),
+				},
+				ApiVersionId: fmt.Sprintf("v%d", j),
+			}
+			// Only take reading for the first version
+			if j == 1 {
+				b.StartTimer()
+			}
```

Comment:
Why do we only measure the first version? Why do we continue creating other versions after that?

Reply:
I wanted to create 3 versions and 3 specs. But the benchmark would add up time of all 3 versions which is how it is reported.

To make sure our reports are accurate we only measure the first pass for revisions and specs but let the scripts create all versions.

Reply:
Right, so let's talk about my second question.

>Why do we continue creating other versions after that?

This is supposed to be benchmarking CreateApiVersion alone. Each iteration of the benchmark loop should only create one ApiVersion. Why are we creating three?

If I'm understanding correctly, BenchmarkCreateApiVersion a benchmark _and_ setup code for BenchmarkCreateApiSpec. Is that correct?

Tests should work correctly regardless of the order they execute in. We should be able to run `go test -bench=Get ./tests/benchmark/...` to successfully measure the performance of every Get method.